modules/javafx.controls/src/test/java/test/javafx/scene/control/ControlUtils.java

```diff
@@ -0,0 +1,190 @@
+        TreeItem<String> root = new TreeItem<String>("");
+        root.setExpanded(true);
+        root.getChildren().setAll(
+            new TreeItem<>(""),
+            new TreeItem<>(""),
+            new TreeItem<>("")
+            );
+
+        TreeTableView<String> t = new TreeTableView<>();
+        t.setRoot(root);
+        t.setShowRoot(false);
+        t.requestFocus();
+        t.getColumns().addAll(
+            createTreeTableColumn("C0"),
+            createTreeTableColumn("C1"),
+            createTreeTableColumn("C2")
+            );
+        return t;
+    }
+
+    /**
+     * Performs a node lookup, returning a TreeTableCell at the given (row, column),
+     * or throws an Error if not found.
+     */
+    public static TreeTableCell getTreeTableCell(TreeTableView t, int row, int column) {
+        TreeTableColumn col = (TreeTableColumn)t.getColumns().get(column);
+
+        for (Node n: t.lookupAll(".tree-table-cell")) {
+            if (n instanceof TreeTableCell c) {
+                if (row == c.getTableRow().getIndex()) {
+                    if (col == c.getTableColumn()) {
+                        return c;
+                    }
+                }
+            }
+        }
+        throw new Error("TreeTableCell not found at " + row + ":" + column);
+    }
+
+    /**
+     * Performs a node lookup, returning a TreeTableRow at the given row,
+     * or throws an Error if not found.
+     */
+    public static TreeTableRow getTreeTableRow(TreeTableView t, int row) {
+        for (Node n: t.lookupAll(".tree-table-row-cell")) {
+            if (n instanceof TreeTableRow c) {
+                if (row == c.getIndex()) {
+                    return c;
+                }
+            }
+        }
+        throw new Error("TreeTableRow not found at " + row);
+    }
+
+    /**
+     * Performs a node lookup, returning a TableCell at the given (row, column)
+     * or throws an Error if not found.
+     */
+    public static TableCell getTableCell(TableView t, int row, int column) {
+        TableColumn col = (TableColumn)t.getColumns().get(column);
+
+        for (Node n: t.lookupAll(".table-cell")) {
+            if (n instanceof TableCell c) {
+                if (row == c.getTableRow().getIndex()) {
+                    if (col == c.getTableColumn()) {
+                        return c;
+                    }
+                }
+            }
+        }
+        throw new Error("TableCell not found at " + row + ":" + column);
+    }
+
+    /**
+     * Performs a node lookup, returning a TableRow at the given row,
+     * or throws an Error if not found.
+     */
+    public static TableRow getTableRow(TableView t, int row) {
+        for (Node n: t.lookupAll(".table-row-cell")) {
+            if (n instanceof TableRow c) {
+                if (row == c.getIndex()) {
+                    return c;
+                }
+            }
+        }
+        throw new Error("TableRow not found at " + row);
+    }
```

Comment:
no review, just a couple of comments:

Generally, should prefer api over lookup - we have shims to access package/protected api. 

In this particular case, the lookup is brittle: it assumes that

- either there is a single row configured with the given index 
- or that it's the first when iterating over the set

With the first being incorrect (corner case, of course :)

    @Test
    public void testLookupLastNotEmptyCell() {
        stageLoader = new StageLoader(table);
        int index = table.getItems().size()- 1;
        Set<Node> tableRows = table.lookupAll(".table-row-cell").stream()
                .filter(n -> n instanceof TableRow<?> && ((TableRow<?>) n).getIndex() == index)
                .collect(Collectors.toSet());
        assertEquals(1, tableRows.size());
    }

we fall back to the second assumption, which is implementation dependent (and unspecified) - it's accidental that we actually do get the row we want.

Having done a bit of digging into VirtualFlowTestUtils (see [JDK-8293356](https://bugs.openjdk.org/browse/JDK-8293356)) I now think it's safe enough to use it (even though it's on the oldish side, not using recently added api, probably needs some polish) - as long as  we keep the scenegraph stable during the test, f.i. by manually adding the table to a stage (either via StageLoader or showing in our own).

Reply:
Is it possible to have two or more rows with the same row index?  i would imagine that will break a lot of things.

I am not sure why you think the lookup is brittle - after all, VirtualFlowTestUtils uses lookup to get a pointer to the VirtualFlow, is it not (line 333)?  Is it because the lookup may not return a cell if it is outside of the current view port area and therefore has not been created (while the .getCell() guarantees to create one)? 

Is this the only objection to the changes in this PR?  I suppose i could change the tests to use VirtualFlowTestUtils, though my preference would be still to test close to the real world scenario, using public APIs (as opposed to getting into internals using some internal test utils), given the fact that I *do* expect these cells to be visible given the dimensions of the table and columns.

Reply:
> Is it possible to have two or more rows with the same row index? i would imagine that will break a lot of things.

don't imagine, verify ;) What happens when you run the test? What happens when you look at the scenegraph (f.i. with ScenicView)? Don't know why they are there but they are - well outside the viewport.

> 
> Is this the only objection to the changes in this PR?

I honestly cannot understand how you possibly could have reached the conclusion that this is the _only_ objection given my comments .. 

Anyway, I'm off here - don't know how to abort a review, but not going to waste more time and nerves on who doesn't seem to listen (that's my personal feeling, of course ;)

Reply:
sorry, what I really meant is - are the code changes ok and we are only discussing updating tests?

my goal here is to fix as many bugs as I possibly can, while listening to the feedback of more experienced team members and addressing their concerns.  it is not always clear when an exchange of ideas ends and marching orders begin, especially in situations where several possibilities exist.  perhaps a clearly annunciated request for specific changes would work?

I feel really uncomfortable with thousands of defects logged against openjfx, and all I want is to help.